src/crypto/muhash.cpp

```diff
@@ -141,98 +130,346 @@ void Num3072::FullReduce()
+        if (eta < 0) {
+            limb_t tmp;
+            eta = -eta;
+            tmp = f; f = g; g = -tmp;
+            tmp = u; u = q; q = -tmp;
+            tmp = v; v = r; r = -tmp;
+        }
+        /* eta is now >= 0. In what follows we're going to cancel out the bottom bits of g. No more
+         * than i can be cancelled out (as we'd be done before that point), and no more than eta+1
+         * can be done as its sign will flip once that happens. */
+        int limit = ((int)eta + 1) > i ? i : ((int)eta + 1);
+        /* m is a mask for the bottom min(limit, 8) bits (our table only supports 8 bits). */
+        limb_t m = (MAX_LIMB >> (LIMB_SIZE - limit)) & 255U;
+        /* Find what multiple of f must be added to g to cancel its bottom min(limit, 8) bits. */
+        limb_t w = (g * NEGINV256[(f >> 1) & 127]) & m;
+        /* Do so. */
+        g += f * w;
+        q += u * w;
+        r += v * w;
+    }
+    out.u = (signed_limb_t)u;
+    out.v = (signed_limb_t)v;
+    out.q = (signed_limb_t)q;
+    out.r = (signed_limb_t)r;
+    return eta;
 }
 
-void Num3072::Multiply(const Num3072& a)
+/** Apply matrix t/2^SIGNED_LIMB_SIZE to vector [d,e], modulo modulus.
+ *
+ * On input and output, d and e are in range 1-2*modulus..modulus-1.
+ */
+inline void UpdateDE(Num3072Signed& d, Num3072Signed& e, const SignedMatrix& t)
 {
-    limb_t c0 = 0, c1 = 0, c2 = 0;
-    Num3072 tmp;
-
-    /* Compute limbs 0..N-2 of this*a into tmp, including one reduction. */
-    for (int j = 0; j < LIMBS - 1; ++j) {
-        limb_t d0 = 0, d1 = 0, d2 = 0;
-        mul(d0, d1, this->limbs[1 + j], a.limbs[LIMBS + j - (1 + j)]);
-        for (int i = 2 + j; i < LIMBS; ++i) muladd3(d0, d1, d2, this->limbs[i], a.limbs[LIMBS + j - i]);
-        mulnadd3(c0, c1, c2, d0, d1, d2, MAX_PRIME_DIFF);
-        for (int i = 0; i < j + 1; ++i) muladd3(c0, c1, c2, this->limbs[i], a.limbs[j - i]);
-        extract3(c0, c1, c2, tmp.limbs[j]);
+    const signed_limb_t u = t.u, v=t.v, q=t.q, r=t.r;
+
+    /* [md,me] start as zero; plus [u,q] if d is negative; plus [v,r] if e is negative. */
+    signed_limb_t sd = d.limbs[SIGNED_LIMBS - 1] >> (LIMB_SIZE - 1);
+    signed_limb_t se = e.limbs[SIGNED_LIMBS - 1] >> (LIMB_SIZE - 1);
+    signed_limb_t md = (u & sd) + (v & se);
+    signed_limb_t me = (q & sd) + (r & se);
+    /* Begin computing t*[d,e]. */
+    signed_limb_t di = d.limbs[0], ei = e.limbs[0];
+    signed_double_limb_t cd = (signed_double_limb_t)u * di + (signed_double_limb_t)v * ei;
+    signed_double_limb_t ce = (signed_double_limb_t)q * di + (signed_double_limb_t)r * ei;
+    /* Correct md,me so that t*[d,e]+modulus*[md,me] has SIGNED_LIMB_SIZE zero bottom bits. */
+    md -= (MODULUS_INVERSE * limb_t(cd) + md) & MAX_SIGNED_LIMB;
+    me -= (MODULUS_INVERSE * limb_t(ce) + me) & MAX_SIGNED_LIMB;
+    /* Update the beginning of computation for t*[d,e]+modulus*[md,me] now md,me are known. */
+    cd -= (signed_double_limb_t)1103717 * md;
```

Comment:
In a26ce628942243fc9848a63bfdfa5e61f5e936f3 "Safegcd based modular inverse for Num3072"

nit: `1103717` appears to be the value of `MAX_PRIME_DIFF`, so I think this can use that variable instead.